pandas/core/dtypes/cast.py

```diff
@@ -1093,14 +1092,11 @@ def astype_nansafe(
         The dtype was a datetime64/timedelta64 dtype, but it had no unit.
     """
     if arr.ndim > 1:
-        # Make sure we are doing non-copy ravel and reshape.
-        flags = arr.flags
-        flat = arr.ravel("K")
+        flat = arr.ravel()
         result = astype_nansafe(flat, dtype, copy=copy, skipna=skipna)
-        order: Literal["C", "F"] = "F" if flags.f_contiguous else "C"
         # error: Item "ExtensionArray" of "Union[ExtensionArray, ndarray]" has no
         # attribute "reshape"
-        return result.reshape(arr.shape, order=order)  # type: ignore[union-attr]
```

Comment:
IIRC the point of the `flags` stuff was to 1) avoid a copy in ravel and b) retain the same contiguity in the returned array.  is the problem that these goals aren't being achieved, or something else?

(im pretty sure i used the same flags-based pattern in the datetimelike arrays, so whatever fix is used here might need to be applied there too)

Reply:
For the behavior regression, what was happening was the following:

1. Data which is not contiguous gets passed in
2. Depending on structure of this data, `arr.ravel("K")` may choose to index with `F` ordering.
3. `order = "F" if flags.f_contiguous else "C"` will give `C` order since the input is not contiguous.
4. So the `reshape` will index with `C` ordering, changing the data ordering

Reply:
so what if we change L1100 to something like `order = "F" if flags.f_contiguous else "C" if flags.c_contiguous else None`?

Reply:
`reshape` doesn't have a concept of `None` -> it can default to `C` if no argument passed, otherwise one of `C`, `F`, `A`. 

To be clear, the `order` argument for `reshape` doesn't have to do with contiguity - it has to do with the order in which elements are read from the input and written to the output (so specifying the order `C` does not mean that the underlying input must be C-contiguous, it means that the input will be read with `C` ordering)

Reply:
Got it, thanks for walking me through this